--- a/arch/mips/ath79/dev-m25p80.c
+++ b/arch/mips/ath79/dev-m25p80.c
@@ -50,7 +50,7 @@ static struct ath79_spi_platform_data at
 void __init ath79_register_m25p80(struct flash_platform_data *pdata)
 {
 	ath79_spi_data.bus_num = 0;
-	ath79_spi_data.num_chipselect = 1;
+	ath79_spi_data.num_chipselect = 3;
 	ath79_spi0_cdata.is_flash = true;
 	ath79_spi_info[0].platform_data = pdata;
 	ath79_register_spi(&ath79_spi_data, ath79_spi_info, 1);
